### PR TITLE
[Feature][BAGEL] Combine multi-branch cfg into a single batch to accelerate inference.

### PR DESCRIPTION
**Pull request by nussejzz**
Feat(bagel):  Add Batched CFG, Timing Logs & Single-Stage Text2Img

## Purpose

- Batched CFG (3-branch & 2-branch): Implements batched Classifier-Free Guidance (CFG) during the denoising phase using packed attention and variable-length Flash Attention (flash_attn_varlen_func). Instead of sequential forward passes for each CFG branch (e.g., conditional, unconditional), sequences are packed into a single batch for a unified LLM forward pass. This significantly reduces overhead and improves inference speed.
- Performance Logging: Adds detailed timing and progress logs (prefill, denoise, vae_decode, per-step latencies) to facilitate granular performance analysis and debugging.
- Bug Fix: Fixes a ZeroDivisionError that occurred when num_timesteps=1 (0 denoising steps), ensuring the pipeline handles edge cases gracefully.

## Test Plan
- Test Environment:  stage1 only. A6000(48GB) gpu_memory_utilization = 0.85 & A100(80GB) gpu_memory_utilization = 0.75
- Test Mode: i2i (3 branches) & t2i (2 branches)
- Verification Script: Run examples/offline_inference/bagel/end2end.py with ByteDance-Seed/BAGEL-7B-MoT.
- Performance Check: Verify the new timing logs in the output (e.g., Denoising: X steps..., Pipeline total: ...) and compare throughput.
- Correctness: Ensure the generated images are visually correct and consistent with the non-batched baseline.
- Edge Case: Run with --num-scheduler-steps 1 to verify the fix for ZeroDivisionError.
- Refactoring: Confirm that text2img requests are correctly routed through the single-stage pipeline.

## Test Result

* The image quality is indistinguishable to the naked eye, and the inference speed is effectively improved, especially the three-branch cfg of I2I, which shows a significant speedup of about 14%. Each denoising step reduces the time by 0.1 seconds. With cfg enabled in the first forty steps, the speedup is reduced by about 4 seconds. *

#### A6000(48GB) gpu_memory_utilization: 0.85 stage1 only
image

#### A100(80GB) gpu_memory_utilization: 0.75 stage1 only
image

### t2I
image

image

image

### ⚠️ Trade-off Note: Memory vs. Speed during CFG Batching

While batching the CFG branches yields a significant speedup, it introduces a transient memory bump. Specifically, using `torch.cat` to merge the KV caches across 2-3 branches means the newly created `merged` cache will temporarily consume 2-3x the memory of a single branch's KV cache. 

**For end-users:** If you are generating images with very large contexts or running on GPUs with limited VRAM, this transient memory spike might lead to Out-Of-Memory (OOM) errors. If you encounter OOMs during the CFG stage, it is highly recommended to slightly lower the `gpu_memory_utilization` parameter to leave more headroom for this merged cache allocation.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please providing the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please pasting the results comparison before and after, or e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by nussejzz** [comment]
PTAL @princepride 😊❤️

File: examples/offline_inference/bagel/end2end.py [hunk 1]
```diff
@@ -81,15 +81,15 @@ def main():
 
     from PIL import Image
 
-    if args.modality == "img2img":
+    if args.modality in ("img2img", "text2img"):
```
Comment (princepride) [inline examples/offline_inference/bagel/end2end.py]: I think we don't need change this file. Maybe you can directly use `OmniDiffusion` test `text2img` task

File: vllm_omni/config/model.py [hunk 2]
```diff
@@ -155,4 +155,4 @@ def __post_init__(
             self.max_model_len = self.get_and_verify_max_len(self.original_max_model_len)
             # Reset sliding_window if needed
             if self.disable_sliding_window:
-                self.hf_text_config.sliding_window = None
+                self.hf_text_config.sliding_window = None
```
Comment (princepride) [inline vllm_omni/config/model.py]: Don't change unrelated code.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 3]
```diff
@@ -27,13 +28,16 @@
     QKVParallelLinear,
     RowParallelLinear,
 )
+from vllm.logger import init_logger
 from vllm.model_executor.layers.vocab_parallel_embedding import VocabParallelEmbedding
 from vllm.model_executor.model_loader.weight_utils import default_weight_loader
 from vllm.transformers_utils.configs.bagel import BagelConfig
+from vllm.vllm_flash_attn import flash_attn_varlen_func
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: We must change it? The original implement will auto use different kernel in different device(GPU/XPU...)

File: vllm_omni/model_executor/stage_configs/bagel.yaml
Comment (princepride) [inline vllm_omni/model_executor/stage_configs/bagel.yaml:1]: Don't change this file.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:1]: Remove all log changes

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:1]: Overall it's good for me, but it seems like a lot of the code has just been rearranged. Could we consider keeping the original code as much as possible? And we can remove log code later.

**Comment by princepride** [comment]
DCO and pre-commit failed.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 4]
```diff
@@ -1283,6 +1287,16 @@ def prepare_vae_latent_cfg(self, curr_kvlens, curr_rope, image_sizes):
...
         return generation_input
 
+    #
+    @staticmethod
+    def _merge_naive_caches(caches: list) -> NaiveCache:
+        """Merge multiple NaiveCache objects by concatenating KV tensors per layer."""
+        merged = NaiveCache(caches[0].num_layers)
+        for layer_idx in range(merged.num_layers):
+            merged.key_cache[layer_idx] = torch.cat([c.key_cache[layer_idx] for c in caches], dim=0)
+            merged.value_cache[layer_idx] = torch.cat([c.value_cache[layer_idx] for c in caches], dim=0)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:1292]: `torch.cat` across all branches here means the merged cache is 2-3× the size of a single branch's KV. For large contexts that could be a pretty meaningful memory bump. Might be worth noting the trade-off in the PR description or at least adding a comment — people tuning `gpu_memory_utilization` would want to know about it.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 5]
```diff
@@ -1414,74 +1484,89 @@ def _forward_flow(
...
                 **extra_inputs,
             )
-            cfg_text_v_t = self.llm2vae(cfg_text_output.packed_query_sequence)
-            cfg_text_v_t = cfg_text_v_t[packed_vae_token_indexes]
 
-        if cfg_img_scale > 1.0:
-            cfg_img_output = self.language_model.forward(
+            # Extract per-branch velocities from batched output
+            all_hidden = output.packed_query_sequence
+            v_t = self.llm2vae(all_hidden[:seq_len])[packed_vae_token_indexes]
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:1498]: The slicing here (`all_hidden[:seq_len]`, `all_hidden[branch_idx * seq_len : ...]`) assumes the language model returns tokens in the exact same packed order as the input. I think that's true for this model, but it's a subtle invariant — if someone changes the attention or adds reordering, this silently produces wrong velocities. A quick shape assert like `assert all_hidden.shape[0] == seq_len * num_branches` would make debugging easier if something ever drifts.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 6]
```diff
@@ -1283,6 +1287,16 @@ def prepare_vae_latent_cfg(self, curr_kvlens, curr_rope, image_sizes):
 
         return generation_input
 
+    #
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Nit: stray `#` comment on the line before the method.

**Review by lishunyang12 — COMMENTED** [review]
A couple things to add on top of princepride's review — mostly about the batched CFG path in the transformer.

**Comment by lishunyang12** [comment]
PTAL

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py
Comment (nussejzz) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1]: Yes, I know, so I only submitted a draft PR to show you the original test environment. The extra progress bar in the logs was only for testing the improvement in inference speed. I will revert everything back to the original.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py (hunk 4 above)
Comment (nussejzz) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1292]: good idea！

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py (hunk 5 above)
Comment (nussejzz) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1498]: High level

Comment (nussejzz) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1498]: In a standard implementation of packed attention + flash attention, the order should indeed remain consistent. However, I believe this assertion is indeed a valuable "defensive programming" tactic.

**Comment by princepride** [comment]
@vllm-omni-reviewer

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py (hunk 4 above)
Comment (lishunyang12) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1292]: Cool, looking forward to seeing the update.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py (hunk 5 above)
Comment (lishunyang12) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1498]: That makes sense — an assertion there sounds like the right tradeoff.

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:1]: I think you can revert this file change, because these time related code only used to print time spending

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 7]
```diff
@@ -1347,24 +1431,28 @@ def generate_image(
...
+        if num_steps > 0:
+            logger.info(
+                "Denoising done: %d steps in %.3fs (avg %.3fs/step)",
+                num_steps,
+                denoise_total,
+                denoise_total / num_steps,
+            )
+        else:
+            logger.info("Denoising done: 0 steps (skipped)")
+
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Remove these log

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 8]
```diff
@@ -1324,7 +1337,78 @@ def generate_image(
...
+            cfg_text_scale,
+            cfg_img_scale,
+            cfg_batched is not None,
+        )
+        torch.cuda.synchronize()
+        denoise_t0 = time.time()
+
         for i, t in enumerate(timesteps):
+            torch.cuda.synchronize()
+            step_t0 = time.time()
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Remove these log.

**Comment by nussejzz** [comment]
Yes, I will revert the changes after I testing the performance improvement with batched cfg + tp @princepride 😋
Finally, I will only change one file: transformer.py

**Comment by princepride** [comment]
@vllm-omni-reviewer

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 9]
```diff
@@ -1414,74 +1459,93 @@ def _forward_flow(
...
-            cfg_img_v_t = self.llm2vae(cfg_img_output.packed_query_sequence)
-            cfg_img_v_t = cfg_img_v_t[packed_vae_token_indexes]
+            v_t = self.llm2vae(output.packed_query_sequence)
+            v_t = v_t[packed_vae_token_indexes]
+            cfg_text_v_t = None
+            cfg_img_v_t = None
 
-        if cfg_text_scale > 1.0:
+        # ── CFG combination ──
+        if cfg_text_scale > 1.0 and cfg_text_v_t is not None:
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: `cfg_text_scale > 1.0` => `use_cfg_text = True` => `cfg_batched != None` => `cfg_text_v_t != None`

So we don't need `cfg_text_v_t is not None`

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 10]
```diff
@@ -1414,74 +1459,93 @@ def _forward_flow(
...
             if cfg_renorm_type == "text_channel":
                 v_t_text_ = cfg_text_v_t + cfg_text_scale * (v_t - cfg_text_v_t)
                 norm_v_t = torch.norm(v_t, dim=-1, keepdim=True)
                 norm_v_t_text_ = torch.norm(v_t_text_, dim=-1, keepdim=True)
                 scale = (norm_v_t / (norm_v_t_text_ + 1e-8)).clamp(min=cfg_renorm_min, max=1.0)
                 v_t_text = v_t_text_ * scale
-                if cfg_img_scale > 1.0:
+                if cfg_img_scale > 1.0 and cfg_img_v_t is not None:
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Same here

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 11]
```diff
@@ -1414,74 +1459,93 @@ def _forward_flow(
...
                     v_t = cfg_img_v_t + cfg_img_scale * (v_t_text - cfg_img_v_t)
                 else:
                     v_t = v_t_text
             else:
                 v_t_text_ = cfg_text_v_t + cfg_text_scale * (v_t - cfg_text_v_t)
 
-                if cfg_img_scale > 1.0:
+                if cfg_img_scale > 1.0 and cfg_img_v_t is not None:
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Same here.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 12]
```diff
@@ -1414,74 +1459,93 @@ def _forward_flow(
...
+            # Extract per-branch velocities from batched output
+            all_hidden = output.packed_query_sequence
+            assert all_hidden.shape[0] == seq_len * num_branches, (
+                f"Expected packed sequence length {seq_len * num_branches}, but got {all_hidden.shape[0]}"
+            )
+
+            v_t = self.llm2vae(all_hidden[:seq_len])[packed_vae_token_indexes]
+
+            cfg_text_v_t = None
+            cfg_img_v_t = None
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Can we initialize `cfg_text_v_t` and `cfg_img_v_t` before `if use_cfg and cfg_batched is not None:`

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 13]
```diff
@@ -1414,74 +1459,93 @@ def _forward_flow(
...
                 update_past_key_values=False,
                 is_causal=False,
                 **extra_inputs,
             )
-            cfg_img_v_t = self.llm2vae(cfg_img_output.packed_query_sequence)
-            cfg_img_v_t = cfg_img_v_t[packed_vae_token_indexes]
+            v_t = self.llm2vae(output.packed_query_sequence)
+            v_t = v_t[packed_vae_token_indexes]
+            cfg_text_v_t = None
+            cfg_img_v_t = None
```
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Same here.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py
Comment (princepride) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:1]: I thought current `if-condition` is too complex to understand, maybe we can only use `if cfg_text_scale` and `if cfg_img_scale` in most case.

Comment (nussejzz) [reply vllm_omni/diffusion/models/bagel/bagel_transformer.py:1]: Thanks for your review! Let me check and edit @princepride

**Comment by princepride** [comment]
> @vllm-omni-reviewer

@hsliuustc0106 Why wasn't it triggered?🤔

**Comment by hsliuustc0106** [comment]
> > @vllm-omni-reviewer
> 
> @hsliuustc0106 Why wasn't it triggered?🤔

@vllm-omni-reviewer

**Comment by princepride** [comment]
@tjtanaa CI failed, PTAL

**Comment by princepride** [comment]
@vllm-omni-reviewer

**Comment by princepride** [comment]
image
@hsliuustc0106 That's weried.
